synthesize/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+use std::collections::HashMap;
+use std::error::Error;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use bootspec::{
+    BootJson, BootSpecPath, SpecialisationDescription, SpecialisationName, SystemConfigurationRoot,
+    JSON_FILENAME, SCHEMA_VERSION,
+};
+
+#[doc(hidden)]
+pub type Result<T, E = Box<dyn Error + Send + Sync + 'static>> = core::result::Result<T, E>;
+
+/// Synthesize a [`BootJson`] struct from the path to a generation.
+///
+/// This is useful when used on generations that do not have a bootspec attached to it.
+pub fn synthesize_schema_from_generation(generation: &Path, out_path: &Path) -> Result<()> {
+    fs::create_dir(&out_path)?;
+    let specialisationdir = out_path.join("specialisation");
+    let mut specialisationdir_created = false;
+
+    let mut toplevelspec = describe_system(&generation)?;
+
+    if let Ok(specialisations) = fs::read_dir(generation.join("specialisation")) {
+        for spec in specialisations.map(|res| res.map(|e| e.path())) {
+            let spec = spec?;
+            let name = spec
+                .file_name()
+                .ok_or("Could not get name of specialisation dir")?
+                .to_str()
+                .ok_or("Specialisation dir name was invalid UTF8")?;
+            let toplevel = fs::canonicalize(generation.join(format!("specialisation/{}", name)))?;
+
+            let mut boot_json_path = toplevel.join(JSON_FILENAME);
+            if !boot_json_path.exists() {
+                if !specialisationdir_created {
+                    fs::create_dir(&specialisationdir)?;
+                    specialisationdir_created = true;
+                }
+                let specname = specialisationdir.join(format!("{}.json", name));
+                let subspec = describe_system(&toplevel)?;
+                let pretty = serde_json::to_string_pretty(&subspec)
+                    .map_err(|e| format!("Failed to make pretty JSON from bootspec:\n{}", e))?;
+
+                fs::write(&specname, pretty).map_err(|e| {
+                    format!("Failed to write JSON to '{}':\n{}", out_path.display(), e)
+                })?;
+                boot_json_path = specname;
+            }
```

Comment:
Remove unneeded mutability (only if it makes logic easier to follow)

```suggestion
            let boot_json_path = toplevel.join(JSON_FILENAME);
            let boot_json_path = match boot_json_path.exists() {
                true => boot_json_path,
                false => {
                    if !specialisationdir_created {
                        fs::create_dir(&specialisationdir)?;
                        specialisationdir_created = true;
                    }

                    let specname = specialisationdir.join(format!("{name}.json"));
                    let subspec = describe_system(&toplevel)?;
                    let pretty = serde_json::to_string_pretty(&subspec)
                        .map_err(|e| format!("Failed to make pretty JSON from bootspec:\n{e}"))?;

                    fs::write(&specname, pretty).map_err(|e| {
                        format!("Failed to write JSON to '{}':\n{e}", out_path.display())
                    })?;

                    specname
                },
            };
```

Reply:
This looks like a good idea, and it is just screaming out to be broken out into smaller functions! Thanks!